### PR TITLE
pack userservice in docker container and expose it Kubernetes togethe…

### PR DESCRIPTION
**Pull request by AndreiMushinsky**

### CODE REVIEW

File: userservice/src/main/resources/db/changelog/changelog-master.sql [hunk 1]
```diff
@@ -14,48 +14,48 @@ create table ADDRESS (
...
     primary key (ID),
-    foreign key (ID) references USER(ID)
+    foreign key (ID) references user(ID)
 );
 
-create table PUPIL (
+create table pupil (
     ID bigint not null,
-    BIRTH_DATE date not null default sysdate(),
+    BIRTH_DATE date not null default '2000-01-01',
```
Comment (xpinjection) [inline userservice/src/main/resources/db/changelog/changelog-master.sql]: Probably default value here is not applicable.

File: userservice/src/main/java/com/kyriba/school/userservice/status/StatusController.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.Collections;
+import java.util.Map;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/status")
+public class StatusController {
```
Comment (xpinjection) [inline userservice/src/main/java/com/kyriba/school/userservice/status/StatusController.java]: Very strange endpoint, you have healthcheck from actuator.

File: userservice/pom.xml [hunk 3]
```diff
@@ -56,38 +61,47 @@
...
+				<version>3.8.0</version>
+				<configuration>
+					<release>11</release>
+				</configuration>
+			</plugin>
+			<plugin>
+				<groupId>org.springframework.boot</groupId>
+				<artifactId>spring-boot-maven-plugin</artifactId>
+			</plugin>
+			<plugin>
```
Comment (xpinjection) [inline userservice/pom.xml]: Usually Docker image is build in dedicated Maven profile.

File: userservice/pom.xml [hunk 4]
```diff
@@ -56,38 +61,47 @@
...
 						</goals>
-						<configuration>
-							<backend>html</backend>
-							<doctype>book</doctype>
-						</configuration>
 					</execution>
 				</executions>
-				<dependencies>
-					<dependency>
-						<groupId>org.springframework.restdocs</groupId>
```
Comment (xpinjection) [inline userservice/pom.xml]: Why did you remove documentation plugins?

File: userservice/pom.xml [hunk 5]
```diff
@@ -10,15 +10,20 @@
 	</parent>
 	<groupId>com.kyriba.school</groupId>
 	<artifactId>userservice</artifactId>
-	<version>0.0.1-SNAPSHOT</version>
+	<version>1.0.0</version>
```
Comment (xpinjection) [inline userservice/pom.xml]: Why did you remove SNAPSHOT from the version?

File: userservice/env/mysql/setup.sh [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
Comment (xpinjection) [inline userservice/env/mysql/setup.sh]: What is the role of this file?

File: userservice/env/kubernetes/userservice-app.yaml [hunk 7]
```diff
@@ -0,0 +1,65 @@
+apiVersion: v1
```
Comment (xpinjection) [inline userservice/env/kubernetes/userservice-app.yaml]: The easiest way to start single service with it's storage is via Docker Compose. Kubernetes could be used at the system level.

File: userservice/src/main/resources/db/changelog/changelog-master.sql (hunk 1 above)
Comment (AndreiMushinsky) [reply userservice/src/main/resources/db/changelog/changelog-master.sql]: I agree. But I like the idea of null safety in my code. What might be appropriate null object for date type?

File: userservice/src/main/java/com/kyriba/school/userservice/status/StatusController.java (hunk 2 above)
Comment (AndreiMushinsky) [reply userservice/src/main/java/com/kyriba/school/userservice/status/StatusController.java]: What's strange about it?
K8s make two types of probe: readiness and liveliness.
Simplistic liveness probe checks if the server is responding. Exactly for that purpose there is this status endpoint that give quick response if my app is not responding. It shouldn't be influenced by an external factor, e.g. liveness probe shouldn’t return a failure when the server can’t connect to the database. If the cause is in the  database  itself,  restarting  the  web  server  container  won't  fix  the  problem. Because the liveness probe will fail again, you’ll end up with the container restarting repeatedly until the database becomes accessible again. Moreover,  the  probes  are  executed  relatively  often  and  are only allowed one second to complete.
A the same time the readiness probe determines whether the specific pod should receive client requests or not. If a pod reports that it’s not ready, it’s removed from the service (not restarted). If the pod then becomes ready again, it’s re-added. So if database is down or file storage is not enough, there is no need to restart a pod, it doesn't fix the problem, we need fix database issue and/or reclaim additional storage, after that a pod is ready again and can serve customer's requests.
So basically that's why /health endpoint is used as readiness probe and /status - as liveliness probe.

File: userservice/pom.xml (hunk 4 above)
Comment (AndreiMushinsky) [reply userservice/pom.xml]: Cause I don't use it :/

File: userservice/pom.xml (hunk 5 above)
Comment (AndreiMushinsky) [reply userservice/pom.xml]: release

File: userservice/env/mysql/setup.sh (hunk 6 above)
Comment (AndreiMushinsky) [reply userservice/env/mysql/setup.sh]: It's a script that starts a docker container with mysql database, nothing special.

File: userservice/env/kubernetes/userservice-app.yaml (hunk 7 above)
Comment (AndreiMushinsky) [reply userservice/env/kubernetes/userservice-app.yaml]: what do you mean?
>Kubernetes could be used at the system level.

Comment (xpinjection) [reply userservice/env/kubernetes/userservice-app.yaml]: It is unusual case to start single service in k8s, that is why people provide easier way to start like Docker compose.

File: userservice/env/mysql/setup.sh (hunk 6 above)
Comment (xpinjection) [reply userservice/env/mysql/setup.sh]: Again it is easier to specify everything in Docker Compose file and then just  run it.

File: userservice/pom.xml (hunk 5 above)
Comment (xpinjection) [reply userservice/pom.xml]: :) but you are still in progress with changes

File: userservice/pom.xml (hunk 4 above)
Comment (xpinjection) [reply userservice/pom.xml]: but it is part of our goal to have self-documented services :)

File: userservice/src/main/java/com/kyriba/school/userservice/status/StatusController.java (hunk 2 above)
Comment (xpinjection) [reply userservice/src/main/java/com/kyriba/school/userservice/status/StatusController.java]: Usually in practice people use /health endpoint for both cases. Take a look at /info endpoint as an option.

File: userservice/src/main/resources/db/changelog/changelog-master.sql (hunk 1 above)
Comment (xpinjection) [reply userservice/src/main/resources/db/changelog/changelog-master.sql]: It is almost impossible to have null safety in the code in Java. All defaults for date type will be dangerous from the data perspective.
